### PR TITLE
enable raw buffer store, so we can do oob store

### PR DESCRIPTION
**Pull request by asleepzzz**
this PR enable oob store for 
fp32,fp16 in xdlop, non xdlops

bf16 only in xdlops due to we do not have non xdlop bf16 support

### CODE REVIEW

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 1]
```diff
@@ -7204,21 +7202,43 @@ struct ThreadwiseCopyV2RewritePattern
...
+                         ifWithinBoundsOp.getResults()[5]});
+        }
       } else {
         if (dataType == b.getF32Type()) {
           b.create<StoreOp>(loc, scalarValue, op.dest(), destLowerIndices);
         } else if (dataType == b.getF16Type()) {
           auto truncValue = b.create<FPTruncOp>(loc, scalarValue, dataType);
           b.create<StoreOp>(loc, truncValue, op.dest(), destLowerIndices);
         } else if (dataType == b.getIntegerType(16)) {
+          // bf16 rsults not correct when in_h & in_w odd
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h]: Add FIXME for bf16

**Review by whchung — CHANGES_REQUESTED** [review]
Add a new unit test as we introduce a new op in this PR

File: mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp [hunk 2]
```diff
@@ -435,6 +435,7 @@ struct MubufStoreOpLowering : ConvertToLLVMPattern {
           loc, interimLLVMValueType, adaptorValue);
       rewriter.replaceOpWithNewOp<ROCDL::MubufStoreOp>(
           op, bitcastedValue, rsrc, vindex, voffset_shift, slc, glc);
+
```
Comment (sjw36) [inline mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp]: remove newline.

File: mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp [hunk 3]
```diff
@@ -444,6 +445,149 @@ struct MubufStoreOpLowering : ConvertToLLVMPattern {
...
+    Value ptrBitcasted =
+        rewriter.create<LLVM::BitcastOp>(loc, LLVMI32x2Type, ptrToInt);
+    Value rsrcUndefTwoItems =
+        rewriter.create<LLVM::UndefOp>(loc, LLVMI32x2Type);
+    Value rsrcFirstTwoItems = rewriter.create<LLVM::ShuffleVectorOp>(
+        loc, ptrBitcasted, rsrcUndefTwoItems,
+        rewriter.getI32ArrayAttr({0, 1, -1, -1}));
+
+    Value rsrcUndef = rewriter.create<LLVM::UndefOp>(loc, LLVMRsrcVectorType);
+    // word 2: fixed as -1 .
```
Comment (sjw36) [inline mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp]: comment doesn't match logic (-1 vs 2)

File: mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp [hunk 4]
```diff
@@ -444,6 +445,149 @@ struct MubufStoreOpLowering : ConvertToLLVMPattern {
...
+          interimValueType =
+              VectorType::get(interimShape, rewriter.getF32Type());
+        Type interimLLVMValueType =
+            typeConverter->convertType(interimValueType);
+
+        Value bitcastedValue = rewriter.create<LLVM::BitcastOp>(
+            loc, interimLLVMValueType, adaptorValue);
+        rewriter.replaceOpWithNewOp<ROCDL::RawbufStoreOp>(
+            op, bitcastedValue, rsrc, voffset_shift, vindex, zeroglcslc);
+      } else { // f16 and i16 (bf16) types scalar value
```
Comment (whchung) [inline mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp:574]: add FIXME

File: mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp (hunk 2 above)
Comment (whchung) [inline mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp]: nit: don't need this newline

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 5]
```diff
@@ -6758,8 +6758,7 @@ struct ThreadwiseCopyRewritePattern
 
         // Store to dest.
         // Issue scalar store ,oob store only support f32 now
-        if (toEmitOOBStoreCheckLogic &&
-            destType.getElementType() == b.getF32Type()) {
+        if (toEmitOOBStoreCheckLogic) {
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:6760]: the comment above needs to be fixed.

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 6]
```diff
@@ -7130,8 +7129,7 @@ struct ThreadwiseCopyV2RewritePattern
       // Store to dest.
       // Issue scalar store.
       // oob store only support f32 now
-      if (toEmitOOBStoreCheckLogic &&
-          destType.getElementType() == b.getF32Type()) {
+      if (toEmitOOBStoreCheckLogic) {
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:7130]: the comment above needs to be fixed.

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 7]
```diff
@@ -7204,21 +7202,43 @@ struct ThreadwiseCopyV2RewritePattern
...
+          auto truncValue = b.create<FPTruncOp>(loc, scalarValue, dataType);
+          b.create<gpu::RawbufStoreOp>(
+              loc, truncValue, op.dest(), ifWithinBoundsOp.getResults()[0],
+              ValueRange{ifWithinBoundsOp.getResults()[1],
+                         ifWithinBoundsOp.getResults()[2],
+                         ifWithinBoundsOp.getResults()[3],
+                         ifWithinBoundsOp.getResults()[4],
+                         ifWithinBoundsOp.getResults()[5]});
+        } else if (dataType == b.getIntegerType(16)) {
+          // bf16 rsults not correct when in_h & in_w odd
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h]: add FIXME.

**Review by sjw36 — APPROVED** [review]
functionality looks good. minor comment cleanup.

**Review by whchung — APPROVED** [review]

**Comment by whchung** [comment]
jenkins: retest this please

**Comment by whchung** [comment]
jenkins: retest this please.

**Comment by whchung** [comment]
jenkins: retest this please.

File: mlir/test/Conversion/GPUToROCDL/rawbuf_store.mlir [hunk 8]
```diff
@@ -0,0 +1,175 @@
+// RUN: mlir-opt %s -convert-gpu-to-rocdl='index-bitwidth=32' | FileCheck %s
+
+gpu.module @mubuf_store {
+  // f32 tests.
+
+  // CHECK-LABEL: func @buffer_store_f32_to_rank_1
+  func @buffer_store_f32_to_rank_1(%value : f32, %dst : memref<128xf32>, %shift : i32, %offset0 : i32) {
+    // CHECK: rocdl.buffer.store %{{.*}}, %{{.*}}, %{{.*}}, %{{.*}}, %{{.*}}, %{{.*}} : f32
+    gpu.buffer_store(%value, %dst, %shift, %offset0) : f32, memref<128xf32>, i32
```
Comment (whchung) [inline mlir/test/Conversion/GPUToROCDL/rawbuf_store.mlir:9]: turns out we are NOT really testing `gpu.raw_buffer_store` and `rocdl.raw.buffer.store` . We are considered lucky that the whole thing works.

**Pull request merged**